libcst/matchers/_matcher_base.py

```diff
@@ -0,0 +1,638 @@
+            return True
+        elif isinstance(matcher, collections.abc.Sequence):
+            # We should assume that this matcher is a sequence to compare. Given
+            # the way we generate match classes, this should be true unless the
+            # match is badly constructed and types were ignored.
+            return _sequence_matches(
+                node,
+                cast(
+                    Sequence[
+                        Union[
+                            BaseMatcherNode,
+                            _BaseWildcardNode,
+                            MatchIfTrue[Callable[..., bool]],
+                            DoNotCareSentinel,
+                        ]
+                    ],
+                    matcher,
+                ),
+            )
+
+        # We exhausted our possibilities, there's no match
+        return False
+
+    # Base case, should match node via matcher. We know the type of node is
+    # correct here because we generate matchers directly off of LibCST nodes,
+    # so the only way it is wrong is if the node was badly constructed and
+    # types were ignored.
+    return _matches(
+        cast(Union[MaybeSentinel, RemovalSentinel, libcst.CSTNode], node),
+        cast(Union[BaseMatcherNode, MatchIfTrue], matcher),
+    )
+
+
+def _node_matches(
+    node: libcst.CSTNode,
+    matcher: Union[
+        BaseMatcherNode,
+        MatchIfTrue[Callable[..., bool]],
+        InverseOf[Union[BaseMatcherNode, MatchIfTrue[Callable[..., bool]]]],
+    ],
+) -> bool:
+    # If this is a InverseOf, then invert the result.
+    if isinstance(matcher, InverseOf):
+        return not _node_matches(node, matcher.matcher)
+
+    # Now, check if this is a lambda matcher.
+    if isinstance(matcher, MatchIfTrue):
+        return matcher.func(node)
+
+    # Now, check that the node and matcher classes are the same.
+    if node.__class__.__name__ != matcher.__class__.__name__:
+        return False
+
+    # Now, check that the children match for each attribute.
+    for field in fields(matcher):
+        if field.name == "_metadata":
+            # We don't care about this field, its a dataclasses implementation detail.
+            continue
+
+        desired = getattr(matcher, field.name)
+        actual = getattr(node, field.name)
+        if not _attribute_matches(actual, desired):
+            return False
+
+    # We didn't find a non-match in the above loop, so it matches!
+    return True
+
+
+def _matches(
+    node: Union[MaybeSentinel, libcst.CSTNode],
+    matcher: Union[
+        BaseMatcherNode,
+        MatchIfTrue[Callable[..., bool]],
+        InverseOf[Union[BaseMatcherNode, MatchIfTrue[Callable[..., bool]]]],
+    ],
+) -> bool:
+    if isinstance(node, MaybeSentinel):
+        # We can't possibly match on a maybe sentinel, so it only matches if
+        # the matcher we have is a InverseOf.
+        return isinstance(matcher, InverseOf)
```

Comment:
This looks like a base case worth a test case.
https://codecov.io/gh/Instagram/LibCST/pull/75/diff?src=pr&el=tree#D6-601